site/content/blog/2019-10-21-localizing-matterpoll/index.md

```diff
@@ -0,0 +1,188 @@
+github: hanzei
+community: hanzei
+---
+
+[Matterpoll](https://github.com/matterpoll/matterpoll) is a plugin that allows users to create polls in Mattermost. Since Mattermost is [localized in 16 different languages](https://docs.mattermost.com/developer/localization.html), it’s optimal that Matterpoll is similarly localized.
+
+Because we rely on contributors to do the translations, we want to make it easy for them to translate new strings and determine if already translated strings needs to be updated because the "source" text changed. On the other hand, Matterpoll only has two maintainer ([@kaakaa](https://github.com/kaakaa) and [me](https://github.com/hanzei)) and no infrastructure of its own to work with. Using a translation server like [Transifex](https://en.wikipedia.org/wiki/Transifex) or [Weblate](https://en.wikipedia.org/wiki/Weblate) is not an option.
+
+The [Mattermost Server](https://github.com/mattermost/mattermost-server) uses the [go-i18n package](https://github.com/nicksnyder/go-i18n). The library is well maintained and very popular, which makes it an attractive tool for this purpose.
+
+This blog provides an outline for developers of how plugins can use the existing framework to support localization.
+
+
+#### Choosing a Version
+
+When [Translations where added to Mattermost](https://github.com/mattermost/mattermost-server/commit/8e404c1dcf820cf767e9d6899e8c1efc7bb5ca96#diff-db85c0ea4d2e69c8abaefa875ba77c51) in 2016, the latest version of go-i18 available was [`v1.4.0`](https://github.com/nicksnyder/go-i18n/releases/tag/v1.4.0). In May of this year [`v2.0.0`](https://github.com/nicksnyder/go-i18n/releases/tag/v2.0.0) was released. V2 has a completely different API and the CLI tool [`goi18n`](https://github.com/nicksnyder/go-i18n#command-goi18n-) has changed significantly. Hence, we had to decide whether to stick with the proven v1 or use the newer v2. 
+
+Let's examine the difference between the two versions. In v1, translation strings are defined in the translation files; for example, `en-US.all.json` would contain:
+```json
+  {
+    "id": "settings_title",
+    "translation": "Settings"
+  }
+```
+These translations would then be accessible in the Go code with a translation function (`T("settings_title")`). This leads to a loose coupling between the translations string in the translation file and in the code. Unfortunately developers tend to forget to add a translation string or misspell the id.
+
+V2 fixed this by defining the translations string in the code and automatically extracting it into the translation file. This greatly improves the developer experience. In addition, there are now **two** translation files for every language. For example, German has `active.de.json`, which contains already translated strings, and `translate.de.json`, which contains strings that have to be translated. They are either newly-added strings or strings where the text changed. `goi18n` helps with auto-populating these files.
```

Comment:
```suggestion
This is fixed in v2 by defining the translations string in the code and automatically extracting it into the translation file. This greatly improves the developer experience. In addition, there are now **two** translation files for every language. For example, German has `active.de.json`, which contains already translated strings, and `translate.de.json`, which contains strings that have to be translated. They are either newly-added strings or strings where the text changed - and `goi18n` helps with auto-populating these files.
```

Reply:
I would like to keep it as it is to have an emphasis on `v2`